odl/contrib/pytorch/modules/haarpsi.py

```diff
@@ -0,0 +1,340 @@
+class SimilarityWeightMap(nn.Module):
+
+    """Pointwise weight map for computation of the HaarPSI FOM.
+
+    For input images :math:`f_1, f_2`, this module computes
+
+    .. math::
+        \mathrm{W}_{f_1, f_2}^{(k)}(x) =
+        \max \\left\{
+        \\left|g_3^{(k)} \\ast f_1 \\right|(x),
+        \\left|g_3^{(k)} \\ast f_2 \\right|(x)
+        \\right\},
+
+    see `[Rei+2016] <https://arxiv.org/abs/1607.06140>`_ equations (11)
+    and (13).
+    Here, the superscript :math:`(k)` refers to the axis (0 or 1) in which
+    edge features are compared, and the 3rd-level Haar wavelet filters
+    :math:`g_3^{(k)}` are high-pass in the axis :math:`k` and low-pass in
+    the other axis, respectively.
+
+    References
+    ----------
+    [Rei+2016] Reisenhofer, R, Bosse, S, Kutyniok, G, and Wiegand, T.
+    *A Haar Wavelet-Based Perceptual Similarity Index for Image Quality
+    Assessment*. arXiv:1607.06140 [cs], Jul. 2016.
+    """
+
+    def __init__(self, axis):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        axis : {0, 1}
+            The axis :math:`k` along which high-pass filters should be
+            applied.
+        """
+        super(SimilarityWeightMap, self).__init__()
+        assert axis in (0, 1)
+        self.axis = int(axis)
+
+    def forward(self, x, y):
+        # Stack input to do them as a batch, add empty in_channels dim
+        xy = torch.stack([x, y], dim=0)[:, None, ...]
+
+        # Init 1d filters for level 3. Since pytorch computes
+        # correlation, not convolution, we flip the filters.
+        filt_lo_lvl1 = np.array([np.sqrt(2), np.sqrt(2)], dtype='float32')
+        filt_hi_lvl1 = np.array([np.sqrt(2), -np.sqrt(2)], dtype='float32')
+        filt_lo_lvl3 = np.repeat(filt_lo_lvl1, 4)
+        filt_hi_lvl3 = np.repeat(filt_hi_lvl1, 4)
+
+        # Make 2D filters by outer products and add empty dims for
+        # out_channels and in_channels
+        if self.axis == 0:
+            # Horizontal high-pass
+            f_l3_arr = np.multiply.outer(filt_hi_lvl3, filt_lo_lvl3)
+            f_l3 = autograd.Variable(
+                torch.from_numpy(f_l3_arr[None, None, ...]))
+
+        else:
+            # Vertical high-pass
+            f_l3_arr = np.multiply.outer(filt_lo_lvl3, filt_hi_lvl3)
+            f_l3 = autograd.Variable(
+                torch.from_numpy(f_l3_arr[None, None, ...]))
+
+        # Do the convolution. Padding must be `kernel_size - 1` to
+        # compute the convolution using all input values. The padding
+        # will be distributed evenly to the "left" and the "right", with
+        # odd paddings adding one extra to the right.
+        conv_l3 = nn.functional.conv2d(xy, f_l3, padding=7)
+
+        conv_x_l3 = conv_l3[0, 0, 3:-4, 3:-4]
+        conv_y_l3 = conv_l3[1, 0, 3:-4, 3:-4]
+
+        # Return maximum of the absolute values
+        return torch.max(torch.abs(conv_x_l3), torch.abs(conv_y_l3))
+
+
+class HaarPSI(nn.Module):
```

Comment:
While I really like this feature. I'm not sure if we should have it in ODL.

One option would frankly be if we (aka, @kohr-h, @adler-j and whoever wants in) created a new library specifically for stuff like this that we can share. One option would be to "steal" the stuff from the adler library as a "base" for the tensorflow stuff and then add stuff like this for torch. By import "gating" we could easily have it all in one library.

Reply:
Sounds good, a deep learning addon library. I don't mind not including the stuff in ODL. I'm also all for hosting something together, since we can do reviews then 😛